### DISCUSSION TITLE
BLE Repeater for GoPros

### DISCUSSION THREAD

**Discussion by sepp89117** [h2zero/NimBLE-Arduino#736] [Q&A] [answered]
Hi, 
I tried to develop a repeater for GoPro BLE communication. 
I have already encountered several obstacles. The first problem is that I can't start services and the server's advertising when I've connected to a GoPro. Alternatively, I now start services and advertising in the setup before a GoPro is connected. 
Now comes the second problem. In the code, I read the data from the GoPro (client) in the server callback onRead to forward it to the peer. Reading the GoPro characteristics leads to a deadlock. 
Is my project technically possible? How can I solve it? I don't necessarily need this application. I thought it was just an interesting programming task.

**Comment by h2zero** [comment, marked as answer]
> I have already encountered several obstacles. The first problem is that I can't start services and the server's advertising when I've connected to a GoPro. Alternatively, I now start services and advertising in the setup before a GoPro is connected.

This is a limitation of the stack unfortunately. You cannot change the state of the GATT server while there are any open connections.

> Now comes the second problem. In the code, I read the data from the GoPro (client) in the server callback onRead to forward it to the peer. Reading the GoPro characteristics leads to a deadlock.

This happens because the stack is waiting for you to return from the callback and is stuck until you do. The call to read is queued but until you return it cannot be processed.

So options are limited, the best you might do is read from the GoPro at regular intervals and respond with the latest data when your server is read from.

**Reply by sepp89117** [reply]
Thanks for your quick reply! Ka, I thought so. There are many characteristics and reading out beforehand to have the answer ready in time is not a nice solution. 
I'll think about it again. 
Thank you

**Comment by mhaberler** [comment]
I'm trying something similar - replaying BLE packets remotely at the frame level (unidirectional is ok for a start)

it is possible if clunky with advertisments - set adv data, start advertising, stop, repeat

however I feel I am using the wrong API - too high level if you will

I guess the better approach would be to use a frame-level API if there is such a thing, without any of the stack's state management

what I'd be looking for is the transmit equivalent of [nRF Sniffer](https://www.nordicsemi.com/Products/Development-tools/nRF-Sniffer-for-Bluetooth-LE)

use case is remote development for a very-hard-to-move BLE sensor installation: a level sensor in a huge tank

thankful for any hints, Michael

**Reply by sepp89117** [reply]
Hi Michael, I have to admit that I don't quite understand what you're saying. If you want to provide data via advertising, have you ever heard of beacon?

**Reply by mhaberler** [reply]
yes, that's what I meant with 'set adv data, start advertising, stop, repeat' - beaconing

**Comment by sepp89117** [comment]
I was able to create a working repeater. [GoPro_BLE_Repeater](https://github.com/sepp89117/GoPro_BLE_Repeater)
I start the server and advertising immediately before connecting the GoPro. This is acceptable.
Fortunately, the readable characteristics are used little, and if so, only at the beginning. So I read all readable characteristics immediately after the connection with a GoPro is established and read them repeatedly one by one when a server characteristic is read.

Thanks again for your support!

**Reply by h2zero** [reply]
Nice! Glad you were able to make it work!
